### PR TITLE
Implement getting and setting user avatar_url (#3 & #11)

### PR DESCRIPTION
**Pull request by christarazi**
This PR depends on mujx/matrix-structs#7 first.

### CODE REVIEW

File: tests/client_api.cpp [hunk 1]
```diff
@@ -83,6 +83,54 @@ TEST(ClientAPI, LoginWrongUsername)
...
+                ASSERT_FALSE(err);
+
+                auto const alice_id   = res.user_id;
+                auto const avatar_url = "mxc://matrix.org/wefh34uihSDRGhw34";
+
+                validate_login(alice_id.toString(), res);
+
+                alice->set_avatar_url(avatar_url, [](ErrType err) { ASSERT_FALSE(err); });
+
+                alice->download_user_avatar(
```
Comment (mujx) [inline tests/client_api.cpp]: All the api calls are async, so the order of the callbacks here is not guaranteed. You could call `download_user_avatar` inside `set_avatar_url`'s callback.

**Comment by christarazi** [comment]
Ah yes, silly mistake. I'll have this fixed later tonight when I find some time. Good catch!

On March 7, 2018 11:14:07 PM PST, mujx <notifications@github.com> wrote:
>mujx commented on this pull request.
>
>
>
>> +
>+TEST(ClientAPI, RealUserAvatar)
>+{
>+        auto alice = std::make_shared<Client>("localhost");
>+
>+        alice->login("alice", "secret", [alice](const
>mtx::responses::Login &res, ErrType err) {
>+                ASSERT_FALSE(err);
>+
>+                auto const alice_id   = res.user_id;
>+                auto const avatar_url =
>"mxc://matrix.org/wefh34uihSDRGhw34";
>+
>+                validate_login(alice_id.toString(), res);
>+
>+                alice->set_avatar_url(avatar_url, [](ErrType err) {
>ASSERT_FALSE(err); });
>+
>+                alice->download_user_avatar(
>
>All the api calls are async, so the order of the callbacks here is not
>guaranteed. You could call `download_user_avatar` inside
>`set_avatar_url`'s callback.
>
>-- 
>You are receiving this because you authored the thread.
>Reply to this email directly or view it on GitHub:
>https://github.com/mujx/mtxclient/pull/19#pullrequestreview-102205427

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

File: src/client.cpp [hunk 2]
```diff
@@ -394,3 +404,14 @@ Client::download(const std::string &server,
                   callback(res, content_type, original_filename, err);
           });
 }
+
+void
+Client::download_user_avatar(
+  const mtx::identifiers::User &user_id,
+  std::function<void(const mtx::responses::Profile &, RequestErr)> callback)
+{
+        get<mtx::responses::Profile>("/client/r0/profile/" + user_id.toString(),
```
Comment (mujx) [inline src/client.cpp]: This is retrieving the whole profile. So better rename this method to `get_profile`.

**Comment by christarazi** [comment]
Yes you're correct. I didn't notice in the spec there were specific endpoints to retrieve the `avatar_url`. I might as well rename this method as you mentioned and also provide a specific endpoint just for the `avatar_url`. I'll also update the tests accordingly. 

On March 8, 2018 10:43:34 PM PST, mujx <notifications@github.com> wrote:
>mujx commented on this pull request.
>
>
>
>> @@ -394,3 +404,14 @@ Client::download(const std::string &server,
>                   callback(res, content_type, original_filename, err);
>           });
> }
>+
>+void
>+Client::download_user_avatar(
>+  const mtx::identifiers::User &user_id,
>+  std::function<void(const mtx::responses::Profile &, RequestErr)>
>callback)
>+{
>+        get<mtx::responses::Profile>("/client/r0/profile/" +
>user_id.toString(),
>
>This is retrieving the whole profile. So better rename this method to
>`get_profile`.
>
>-- 
>You are receiving this because you authored the thread.
>Reply to this email directly or view it on GitHub:
>https://github.com/mujx/mtxclient/pull/19#pullrequestreview-102556346

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by mujx** [comment]
Yeah, either way works.

**Comment by christarazi** [comment]
Commit cc56fa0 depends on mujx/matrix-structs#9 first.

**Pull request merged**